File: frontend/lib/api/linkedin.ts
/**
 * LinkedIn API Client
 *
 * TypeScript client for interacting with the LinkedIn integration API.
 */

// ==================== Types ====================

export type ConnectionStatus =
  | 'not_connected'
  | 'pending_sent'
  | 'pending_received'
  | 'connected'
  | 'following';

export type OutreachType =
  | 'connection_request'
  | 'inmail'
  | 'message'
  | 'comment'
  | 'like'
  | 'share'
  | 'profile_view';

export type OutreachStatus =
  | 'pending'
  | 'sent'
  | 'delivered'
  | 'read'
  | 'replied'
  | 'accepted'
  | 'declined'
  | 'expired';

export type ActivityType =
  | 'post'
  | 'article'
  | 'reaction'
  | 'comment'
  | 'share'
  | 'job_change'
  | 'promotion'
  | 'work_anniversary'
  | 'new_connection';

export interface LinkedInExperience {
  title: string;
  company_name: string;
  company_linkedin_url?: string;
  location?: string;
  start_date?: string;
  end_date?: string;
  is_current: boolean;
  description?: string;
}

export interface LinkedInEducation {
  school_name: string;
  school_linkedin_url?: string;
  degree?: string;
  field_of_study?: string;
  start_year?: number;
  end_year?: number;
  description?: string;
}

export interface LinkedInSkill {
  name: string;
  endorsement_count: number;
}

export interface LinkedInProfile {
  linkedin_id?: string;
  linkedin_url: string;
  first_name: string;
  last_name: string;
  headline?: string;
  summary?: string;
  location?: string;
  country?: string;
  industry?: string;
  profile_picture_url?: string;
  banner_image_url?: string;
  current_title?: string;
  current_company?: string;
  current_company_linkedin_url?: string;
  email?: string;
  phone?: string;
  website?: string;
  twitter_handle?: string;
  connections_count?: number;
  followers_count?: number;
  experiences: LinkedInExperience[];
  education: LinkedInEducation[];
  skills: LinkedInSkill[];
  languages: string[];
  is_open_to_work: boolean;
  is_hiring: boolean;
  is_creator: boolean;
  connection_status: ConnectionStatus;
  last_enriched_at?: string;
  enrichment_source?: string;
}

export interface LinkedInProfileSummary {
  linkedin_url: string;
  first_name: string;
  last_name: string;
  headline?: string;
  current_company?: string;
  location?: string;
  profile_picture_url?: string;
  connection_status: ConnectionStatus;
}

export interface LinkedInCompany {
  linkedin_id?: string;
  linkedin_url: string;
  name: string;
  tagline?: string;
  description?: string;
  website?: string;
  industry?: string;
  company_size?: string;
  employee_count?: number;
  headquarters_location?: string;
  headquarters_city?: string;
  headquarters_country?: string;
  founded_year?: number;
  company_type?: string;
  specialties: string[];
  logo_url?: string;
  followers_count?: number;
  last_enriched_at?: string;
  enrichment_source?: string;
}

export interface OutreachActivity {
  id: string;
  prospect_linkedin_url: string;
  prospect_name?: string;
  outreach_type: OutreachType;
  status: OutreachStatus;
  message_content?: string;
  subject?: string;
  created_at: string;
  sent_at?: string;
  delivered_at?: string;
  read_at?: string;
  replied_at?: string;
  response_content?: string;
  campaign_id?: string;
  sequence_step?: number;
  is_sales_navigator: boolean;
}

export interface OutreachCampaign {
  id: string;
  name: string;
  description?: string;
  is_active: boolean;
  start_date?: string;
  end_date?: string;
  target_profiles: string[];
  message_templates: string[];
  total_prospects: number;
  sent_count: number;
  delivered_count: number;
  read_count: number;
  replied_count: number;
  accepted_count: number;
  created_at: string;
  updated_at?: string;
}

export interface LinkedInActivity {
  id: string;
  profile_linkedin_url: string;
  activity_type: ActivityType;
  activity_url?: string;
  content_text?: string;
  likes_count: number;
  comments_count: number;
  shares_count: number;
  old_title?: string;
  old_company?: string;
  new_title?: string;
  new_company?: string;
  activity_date: string;
  discovered_at: string;
}

export interface ConnectionRecord {
  id: string;
  prospect_linkedin_url: string;
  prospect_name?: string;
  previous_status: ConnectionStatus;
  new_status: ConnectionStatus;
  changed_at: string;
  connection_note?: string;
}

export interface EnrichmentResponse {
  success: boolean;
  profile?: LinkedInProfile;
  company?: LinkedInCompany;
  error_message?: string;
  cached: boolean;
  enrichment_source?: string;
}

export interface BulkEnrichmentResponse {
  total_requested: number;
  successful: number;
  failed: number;
  results: EnrichmentResponse[];
}

export interface ProfileMatchResponse {
  matched: boolean;
  confidence_score: number;
  prospect_id?: string;
  profile?: LinkedInProfileSummary;
  match_reasons: string[];
}

export interface OutreachAnalytics {
  total_outreach: number;
  by_type: Record<string, number>;
  by_status: Record<string, number>;
  reply_rate: number;
  acceptance_rate: number;
  period_days: number;
}

// ==================== API Client ====================

const API_BASE = process.env.NEXT_PUBLIC_API_URL || '/api';

async function fetchAPI<T>(
  endpoint: string,
  options: RequestInit = {}
): Promise<T> {
  const url = `${API_BASE}/linkedin${endpoint}`;

  const response = await fetch(url, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      ...options.headers,
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || `API error: ${response.status}`);
  }

  return response.json();
}

// ==================== Profile Enrichment ====================

export async function enrichProfile(
  linkedinUrl: string,
  options: {
    forceRefresh?: boolean;
    includeExperiences?: boolean;
    includeEducation?: boolean;
    includeSkills?: boolean;
  } = {}
): Promise<EnrichmentResponse> {
  return fetchAPI<EnrichmentResponse>('/profiles/enrich', {
    method: 'POST',
    body: JSON.stringify({
      linkedin_url: linkedinUrl,
      force_refresh: options.forceRefresh ?? false,
      include_experiences: options.includeExperiences ?? true,
      include_education: options.includeEducation ?? true,
      include_skills: options.includeSkills ?? true,
    }),
  });
}

export async function bulkEnrichProfiles(
  linkedinUrls: string[],
  forceRefresh = false
): Promise<BulkEnrichmentResponse> {
  return fetchAPI<BulkEnrichmentResponse>('/profiles/enrich/bulk', {
    method: 'POST',
    body: JSON.stringify({
      linkedin_urls: linkedinUrls,
      force_refresh: forceRefresh,
    }),
  });
}

export async function getProfile(
  username: string,
  forceRefresh = false
): Promise<EnrichmentResponse> {
  const params = new URLSearchParams();
  if (forceRefresh) params.set('force_refresh', 'true');
  return fetchAPI<EnrichmentResponse>(
    `/profiles/${username}?${params.toString()}`
  );
}

// ==================== Company Enrichment ====================

export async function enrichCompany(
  linkedinUrl: string,
  options: {
    forceRefresh?: boolean;
    includeKeyEmployees?: boolean;
  } = {}
): Promise<EnrichmentResponse> {
  return fetchAPI<EnrichmentResponse>('/companies/enrich', {
    method: 'POST',
    body: JSON.stringify({
      linkedin_url: linkedinUrl,
      force_refresh: options.forceRefresh ?? false,
      include_key_employees: options.includeKeyEmployees ?? false,
    }),
  });
}

export async function getCompany(
  slug: string,
  forceRefresh = false
): Promise<EnrichmentResponse> {
  const params = new URLSearchParams();
  if (forceRefresh) params.set('force_refresh', 'true');
  return fetchAPI<EnrichmentResponse>(
    `/companies/${slug}?${params.toString()}`
  );
}

// ==================== Outreach Tracking ====================

export async function trackOutreach(data: {
  prospectLinkedinUrl: string;
  outreachType: OutreachType;
  messageContent?: string;
  subject?: string;
  campaignId?: string;
  isSalesNavigator?: boolean;
}): Promise<OutreachActivity> {
  return fetchAPI<OutreachActivity>('/outreach/track', {
    method: 'POST',
    body: JSON.stringify({
      prospect_linkedin_url: data.prospectLinkedinUrl,
      outreach_type: data.outreachType,
      message_content: data.messageContent,
      subject: data.subject,
      campaign_id: data.campaignId,
      is_sales_navigator: data.isSalesNavigator ?? false,
    }),
  });
}

export async function updateOutreachStatus(
  activityId: string,
  status: OutreachStatus,
  responseContent?: string
): Promise<OutreachActivity> {
  return fetchAPI<OutreachActivity>(`/outreach/${activityId}/status`, {
    method: 'PATCH',
    body: JSON.stringify({
      status,
      response_content: responseContent,
    }),
  });
}

export async function listOutreachActivities(filters: {
  prospectLinkedinUrl?: string;
  campaignId?: string;
  outreachType?: OutreachType;
  limit?: number;
} = {}): Promise<OutreachActivity[]> {
  const params = new URLSearchParams();
  if (filters.prospectLinkedinUrl)
    params.set('prospect_linkedin_url', filters.prospectLinkedinUrl);
  if (filters.campaignId) params.set('campaign_id', filters.campaignId);
  if (filters.outreachType) params.set('outreach_type', filters.outreachType);
  if (filters.limit) params.set('limit', filters.limit.toString());
  return fetchAPI<OutreachActivity[]>(`/outreach?${params.toString()}`);
}

export async function getOutreachAnalytics(
  campaignId?: string,
  days = 30
): Promise<OutreachAnalytics> {
  const params = new URLSearchParams();
  if (campaignId) params.set('campaign_id', campaignId);
  params.set('days', days.toString());
  return fetchAPI<OutreachAnalytics>(
    `/outreach/analytics?${params.toString()}`
  );
}

// ==================== Connection Tracking ====================

export async function updateConnectionStatus(
  prospectLinkedinUrl: string,
  status: ConnectionStatus,
  note?: string
): Promise<ConnectionRecord> {
  return fetchAPI<ConnectionRecord>('/connections/status', {
    method: 'POST',
    body: JSON.stringify({
      prospect_linkedin_url: prospectLinkedinUrl,
      status,
      note,
    }),
  });
}

export async function getConnectionStatus(
  prospectLinkedinUrl: string
): Promise<{ status: ConnectionStatus }> {
  const params = new URLSearchParams({
    prospect_linkedin_url: prospectLinkedinUrl,
  });
  return fetchAPI<{ status: ConnectionStatus }>(
    `/connections/status?${params.toString()}`
  );
}

export async function getConnectionHistory(
  prospectLinkedinUrl: string
): Promise<ConnectionRecord[]> {
  const params = new URLSearchParams({
    prospect_linkedin_url: prospectLinkedinUrl,
  });
  return fetchAPI<ConnectionRecord[]>(
    `/connections/history?${params.toString()}`
  );
}

// ==================== Campaign Management ====================

export async function createCampaign(data: {
  name: string;
  description?: string;
  targetProfiles?: string[];
  messageTemplates?: string[];
}): Promise<OutreachCampaign> {
  return fetchAPI<OutreachCampaign>('/campaigns', {
    method: 'POST',
    body: JSON.stringify({
      name: data.name,
      description: data.description,
      target_profiles: data.targetProfiles,
      message_templates: data.messageTemplates,
    }),
  });
}

export async function listCampaigns(
  activeOnly = false
): Promise<OutreachCampaign[]> {
  const params = new URLSearchParams();
  if (activeOnly) params.set('active_only', 'true');
  return fetchAPI<OutreachCampaign[]>(`/campaigns?${params.toString()}`);
}

export async function getCampaign(campaignId: string): Promise<OutreachCampaign> {
  return fetchAPI<OutreachCampaign>(`/campaigns/${campaignId}`);
}

export async function updateCampaign(
  campaignId: string,
  updates: {
    name?: string;
    description?: string;
    isActive?: boolean;
  }
): Promise<OutreachCampaign> {
  const params = new URLSearchParams();
  if (updates.name) params.set('name', updates.name);
  if (updates.description) params.set('description', updates.description);
  if (updates.isActive !== undefined)
    params.set('is_active', updates.isActive.toString());
  return fetchAPI<OutreachCampaign>(
    `/campaigns/${campaignId}?${params.toString()}`,
    { method: 'PATCH' }
  );
}

// ==================== Activity Monitoring ====================

export async function recordActivity(data: {
  profileLinkedinUrl: string;
  activityType: ActivityType;
  contentText?: string;
  activityUrl?: string;
  activityDate?: string;
  oldTitle?: string;
  oldCompany?: string;
  newTitle?: string;
  newCompany?: string;
}): Promise<LinkedInActivity> {
  return fetchAPI<LinkedInActivity>('/activities', {
    method: 'POST',
    body: JSON.stringify({
      profile_linkedin_url: data.profileLinkedinUrl,
      activity_type: data.activityType,
      content_text: data.contentText,
      activity_url: data.activityUrl,
      activity_date: data.activityDate,
      old_title: data.oldTitle,
      old_company: data.oldCompany,
      new_title: data.newTitle,
      new_company: data.newCompany,
    }),
  });
}

export async function listActivities(filters: {
  prospectLinkedinUrl: string;
  activityType?: ActivityType;
  sinceDays?: number;
  limit?: number;
}): Promise<LinkedInActivity[]> {
  const params = new URLSearchParams({
    prospect_linkedin_url: filters.prospectLinkedinUrl,
  });
  if (filters.activityType) params.set('activity_type', filters.activityType);
  if (filters.sinceDays) params.set('since_days', filters.sinceDays.toString());
  if (filters.limit) params.set('limit', filters.limit.toString());
  return fetchAPI<LinkedInActivity[]>(`/activities?${params.toString()}`);
}

export async function getJobChanges(
  sinceDays = 30,
  limit = 50
): Promise<LinkedInActivity[]> {
  const params = new URLSearchParams({
    since_days: sinceDays.toString(),
    limit: limit.toString(),
  });
  return fetchAPI<LinkedInActivity[]>(
    `/activities/job-changes?${params.toString()}`
  );
}

// ==================== Profile Matching ====================

export async function matchProfileToProspect(data: {
  linkedinUrl: string;
  prospectId?: string;
  email?: string;
  firstName?: string;
  lastName?: string;
  company?: string;
}): Promise<ProfileMatchResponse> {
  return fetchAPI<ProfileMatchResponse>('/profiles/match', {
    method: 'POST',
    body: JSON.stringify({
      linkedin_url: data.linkedinUrl,
      prospect_id: data.prospectId,
      email: data.email,
      first_name: data.firstName,
      last_name: data.lastName,
      company: data.company,
    }),
  });
}

// ==================== Search ====================

export async function searchProfiles(filters: {
  query?: string;
  firstName?: string;
  lastName?: string;
  company?: string;
  title?: string;
  location?: string;
  limit?: number;
}): Promise<LinkedInProfileSummary[]> {
  const params = new URLSearchParams();
  if (filters.query) params.set('query', filters.query);
  if (filters.firstName) params.set('first_name', filters.firstName);
  if (filters.lastName) params.set('last_name', filters.lastName);
  if (filters.company) params.set('company', filters.company);
  if (filters.title) params.set('title', filters.title);
  if (filters.location) params.set('location', filters.location);
  if (filters.limit) params.set('limit', filters.limit.toString());
  return fetchAPI<LinkedInProfileSummary[]>(
    `/search/profiles?${params.toString()}`
  );
}

// ==================== Health Check ====================

export async function checkHealth(): Promise<{
  status: string;
  service: string;
  rate_limit: Record<string, any>;
  cache_enabled: boolean;
  enrichment_provider: string;
}> {
  return fetchAPI('/health');
}

File: backend/app/models/linkedin.py
"""
LinkedIn Models and Pydantic Schemas

This module defines the data models for LinkedIn integration including:
- Profile data structures
- Company data structures
- Connection tracking
- Outreach activity tracking
"""

from datetime import datetime
from enum import Enum
from typing import Optional, List
from pydantic import BaseModel, Field, HttpUrl, field_validator
import re


# Enums
class ConnectionStatus(str, Enum):
    """LinkedIn connection status between user and prospect"""
    NOT_CONNECTED = "not_connected"
    PENDING_SENT = "pending_sent"
    PENDING_RECEIVED = "pending_received"
    CONNECTED = "connected"
    FOLLOWING = "following"


class OutreachType(str, Enum):
    """Types of LinkedIn outreach activities"""
    CONNECTION_REQUEST = "connection_request"
    INMAIL = "inmail"
    MESSAGE = "message"
    COMMENT = "comment"
    LIKE = "like"
    SHARE = "share"
    PROFILE_VIEW = "profile_view"


class OutreachStatus(str, Enum):
    """Status of outreach activities"""
    PENDING = "pending"
    SENT = "sent"
    DELIVERED = "delivered"
    READ = "read"
    REPLIED = "replied"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    EXPIRED = "expired"


class CompanySize(str, Enum):
    """Company size categories"""
    SELF_EMPLOYED = "1"
    SMALL_2_10 = "2-10"
    SMALL_11_50 = "11-50"
    MEDIUM_51_200 = "51-200"
    MEDIUM_201_500 = "201-500"
    LARGE_501_1000 = "501-1000"
    LARGE_1001_5000 = "1001-5000"
    ENTERPRISE_5001_10000 = "5001-10000"
    ENTERPRISE_10000_PLUS = "10000+"


class ActivityType(str, Enum):
    """Types of LinkedIn activities to monitor"""
    POST = "post"
    ARTICLE = "article"
    REACTION = "reaction"
    COMMENT = "comment"
    SHARE = "share"
    JOB_CHANGE = "job_change"
    PROMOTION = "promotion"
    WORK_ANNIVERSARY = "work_anniversary"
    NEW_CONNECTION = "new_connection"


# Base Models
class LinkedInExperience(BaseModel):
    """Work experience entry from LinkedIn profile"""
    title: str
    company_name: str
    company_linkedin_url: Optional[str] = None
    location: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    is_current: bool = False
    description: Optional[str] = None

    @property
    def duration_months(self) -> Optional[int]:
        """Calculate duration in months"""
        if not self.start_date:
            return None
        end = self.end_date or datetime.now()
        return (end.year - self.start_date.year) * 12 + (end.month - self.start_date.month)


class LinkedInEducation(BaseModel):
    """Education entry from LinkedIn profile"""
    school_name: str
    school_linkedin_url: Optional[str] = None
    degree: Optional[str] = None
    field_of_study: Optional[str] = None
    start_year: Optional[int] = None
    end_year: Optional[int] = None
    description: Optional[str] = None
    activities: Optional[str] = None


class LinkedInSkill(BaseModel):
    """Skill entry from LinkedIn profile"""
    name: str
    endorsement_count: int = 0


class LinkedInCertification(BaseModel):
    """Certification entry from LinkedIn profile"""
    name: str
    issuing_organization: str
    issue_date: Optional[datetime] = None
    expiration_date: Optional[datetime] = None
    credential_id: Optional[str] = None
    credential_url: Optional[str] = None


# Profile Models
class LinkedInProfile(BaseModel):
    """Complete LinkedIn profile data"""
    linkedin_id: Optional[str] = None
    linkedin_url: str
    first_name: str
    last_name: str
    headline: Optional[str] = None
    summary: Optional[str] = None
    location: Optional[str] = None
    country: Optional[str] = None
    industry: Optional[str] = None
    profile_picture_url: Optional[str] = None
    banner_image_url: Optional[str] = None

    # Current position
    current_title: Optional[str] = None
    current_company: Optional[str] = None
    current_company_linkedin_url: Optional[str] = None

    # Contact info (if available)
    email: Optional[str] = None
    phone: Optional[str] = None
    website: Optional[str] = None
    twitter_handle: Optional[str] = None

    # Network stats
    connections_count: Optional[int] = None
    followers_count: Optional[int] = None

    # Experience and education
    experiences: List[LinkedInExperience] = Field(default_factory=list)
    education: List[LinkedInEducation] = Field(default_factory=list)
    skills: List[LinkedInSkill] = Field(default_factory=list)
    certifications: List[LinkedInCertification] = Field(default_factory=list)
    languages: List[str] = Field(default_factory=list)

    # Engagement metrics
    is_open_to_work: bool = False
    is_hiring: bool = False
    is_creator: bool = False

    # Connection status with our user
    connection_status: ConnectionStatus = ConnectionStatus.NOT_CONNECTED

    # Metadata
    last_enriched_at: Optional[datetime] = None
    enrichment_source: Optional[str] = None

    @field_validator('linkedin_url')
    @classmethod
    def validate_linkedin_url(cls, v: str) -> str:
        """Normalize and validate LinkedIn URL"""
        return normalize_linkedin_url(v)

    @property
    def full_name(self) -> str:
        return f"{self.first_name} {self.last_name}"

    @property
    def total_experience_years(self) -> float:
        """Calculate total years of experience"""
        total_months = sum(
            exp.duration_months or 0
            for exp in self.experiences
        )
        return round(total_months / 12, 1)


class LinkedInProfileSummary(BaseModel):
    """Condensed profile info for list views"""
    linkedin_url: str
    first_name: str
    last_name: str
    headline: Optional[str] = None
    current_company: Optional[str] = None
    location: Optional[str] = None
    profile_picture_url: Optional[str] = None
    connection_status: ConnectionStatus = ConnectionStatus.NOT_CONNECTED


# Company Models
class LinkedInCompany(BaseModel):
    """LinkedIn company page data"""
    linkedin_id: Optional[str] = None
    linkedin_url: str
    name: str
    tagline: Optional[str] = None
    description: Optional[str] = None
    website: Optional[str] = None
    industry: Optional[str] = None
    company_size: Optional[CompanySize] = None
    employee_count: Optional[int] = None
    employee_count_range: Optional[str] = None

    # Location
    headquarters_location: Optional[str] = None
    headquarters_city: Optional[str] = None
    headquarters_country: Optional[str] = None

    # Company details
    founded_year: Optional[int] = None
    company_type: Optional[str] = None  # Public, Private, Nonprofit, etc.
    specialties: List[str] = Field(default_factory=list)

    # Branding
    logo_url: Optional[str] = None
    banner_url: Optional[str] = None

    # Social metrics
    followers_count: Optional[int] = None

    # Associated pages
    showcase_pages: List[str] = Field(default_factory=list)
    affiliated_companies: List[str] = Field(default_factory=list)

    # Key people (if available)
    key_employees: List[LinkedInProfileSummary] = Field(default_factory=list)

    # Metadata
    last_enriched_at: Optional[datetime] = None
    enrichment_source: Optional[str] = None

    @field_validator('linkedin_url')
    @classmethod
    def validate_linkedin_url(cls, v: str) -> str:
        """Normalize and validate LinkedIn company URL"""
        return normalize_linkedin_company_url(v)


# Outreach Models
class OutreachActivity(BaseModel):
    """Track individual outreach activities"""
    id: Optional[str] = None
    prospect_linkedin_url: str
    prospect_name: Optional[str] = None
    outreach_type: OutreachType
    status: OutreachStatus = OutreachStatus.PENDING

    # Content
    message_content: Optional[str] = None
    subject: Optional[str] = None  # For InMails

    # Timing
    created_at: datetime = Field(default_factory=datetime.now)
    sent_at: Optional[datetime] = None
    delivered_at: Optional[datetime] = None
    read_at: Optional[datetime] = None
    replied_at: Optional[datetime] = None

    # Response tracking
    response_content: Optional[str] = None

    # Campaign association
    campaign_id: Optional[str] = None
    sequence_step: Optional[int] = None

    # Sales Navigator fields
    sales_navigator_activity_id: Optional[str] = None
    is_sales_navigator: bool = False

    # User tracking
    user_id: Optional[str] = None

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class OutreachCampaign(BaseModel):
    """Group related outreach activities into campaigns"""
    id: Optional[str] = None
    name: str
    description: Optional[str] = None

    # Campaign settings
    is_active: bool = True
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None

    # Targeting
    target_profiles: List[str] = Field(default_factory=list)  # LinkedIn URLs

    # Sequence
    message_templates: List[str] = Field(default_factory=list)
    delay_between_steps_days: int = 3

    # Stats
    total_prospects: int = 0
    sent_count: int = 0
    delivered_count: int = 0
    read_count: int = 0
    replied_count: int = 0
    accepted_count: int = 0

    # Metadata
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: Optional[datetime] = None
    user_id: Optional[str] = None

    @property
    def reply_rate(self) -> float:
        """Calculate reply rate percentage"""
        if self.sent_count == 0:
            return 0.0
        return round((self.replied_count / self.sent_count) * 100, 2)

    @property
    def acceptance_rate(self) -> float:
        """Calculate connection acceptance rate"""
        if self.sent_count == 0:
            return 0.0
        return round((self.accepted_count / self.sent_count) * 100, 2)


# Activity Monitoring Models
class LinkedInActivity(BaseModel):
    """Track LinkedIn activities from prospects"""
    id: Optional[str] = None
    profile_linkedin_url: str
    activity_type: ActivityType

    # Content
    activity_url: Optional[str] = None
    content_text: Optional[str] = None
    media_url: Optional[str] = None

    # Engagement metrics
    likes_count: int = 0
    comments_count: int = 0
    shares_count: int = 0

    # For job changes/promotions
    old_title: Optional[str] = None
    old_company: Optional[str] = None
    new_title: Optional[str] = None
    new_company: Optional[str] = None

    # Timing
    activity_date: datetime
    discovered_at: datetime = Field(default_factory=datetime.now)

    # Alert settings
    is_alert_sent: bool = False
    alert_sent_at: Optional[datetime] = None


# Connection Tracking
class ConnectionRecord(BaseModel):
    """Track connection status changes over time"""
    id: Optional[str] = None
    prospect_linkedin_url: str
    prospect_name: Optional[str] = None

    # Status
    previous_status: ConnectionStatus
    new_status: ConnectionStatus
    changed_at: datetime = Field(default_factory=datetime.now)

    # Context
    connection_request_message: Optional[str] = None
    connection_note: Optional[str] = None

    # User tracking
    user_id: Optional[str] = None


# Request/Response Models
class ProfileEnrichmentRequest(BaseModel):
    """Request to enrich a LinkedIn profile"""
    linkedin_url: str
    force_refresh: bool = False
    include_experiences: bool = True
    include_education: bool = True
    include_skills: bool = True
    include_certifications: bool = False

    @field_validator('linkedin_url')
    @classmethod
    def validate_url(cls, v: str) -> str:
        return normalize_linkedin_url(v)


class CompanyEnrichmentRequest(BaseModel):
    """Request to enrich a LinkedIn company"""
    linkedin_url: str
    force_refresh: bool = False
    include_key_employees: bool = False
    key_employee_limit: int = 10

    @field_validator('linkedin_url')
    @classmethod
    def validate_url(cls, v: str) -> str:
        return normalize_linkedin_company_url(v)


class BulkEnrichmentRequest(BaseModel):
    """Request to enrich multiple profiles"""
    linkedin_urls: List[str]
    force_refresh: bool = False

    @field_validator('linkedin_urls')
    @classmethod
    def validate_urls(cls, v: List[str]) -> List[str]:
        return [normalize_linkedin_url(url) for url in v]


class EnrichmentResponse(BaseModel):
    """Response from enrichment operation"""
    success: bool
    profile: Optional[LinkedInProfile] = None
    company: Optional[LinkedInCompany] = None
    error_message: Optional[str] = None
    cached: bool = False
    enrichment_source: Optional[str] = None


class BulkEnrichmentResponse(BaseModel):
    """Response from bulk enrichment"""
    total_requested: int
    successful: int
    failed: int
    results: List[EnrichmentResponse]


class OutreachTrackingRequest(BaseModel):
    """Request to track an outreach activity"""
    prospect_linkedin_url: str
    outreach_type: OutreachType
    message_content: Optional[str] = None
    subject: Optional[str] = None
    campaign_id: Optional[str] = None


class ProfileMatchRequest(BaseModel):
    """Request to match a LinkedIn profile to a prospect"""
    linkedin_url: str
    prospect_id: Optional[str] = None
    email: Optional[str] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    company: Optional[str] = None


class ProfileMatchResponse(BaseModel):
    """Response from profile matching"""
    matched: bool
    confidence_score: float
    prospect_id: Optional[str] = None
    profile: Optional[LinkedInProfileSummary] = None
    match_reasons: List[str] = Field(default_factory=list)


# Utility Functions
def normalize_linkedin_url(url: str) -> str:
    """
    Normalize a LinkedIn profile URL to standard format.
    Handles various LinkedIn URL formats:
    - https://www.linkedin.com/in/username
    - https://linkedin.com/in/username/
    - linkedin.com/in/username
    - /in/username
    """
    if not url:
        raise ValueError("LinkedIn URL cannot be empty")

    # Remove whitespace
    url = url.strip()

    # Extract username from various formats
    patterns = [
        r'(?:https?://)?(?:www\.)?linkedin\.com/in/([^/?\s]+)',
        r'^/in/([^/?\s]+)',
        r'^in/([^/?\s]+)',
    ]

    for pattern in patterns:
        match = re.search(pattern, url, re.IGNORECASE)
        if match:
            username = match.group(1)
            return f"https://www.linkedin.com/in/{username}"

    # If URL doesn't match patterns, assume it's just a username
    if '/' not in url and '.' not in url:
        return f"https://www.linkedin.com/in/{url}"

    raise ValueError(f"Invalid LinkedIn profile URL: {url}")


def normalize_linkedin_company_url(url: str) -> str:
    """
    Normalize a LinkedIn company URL to standard format.
    Handles various LinkedIn company URL formats:
    - https://www.linkedin.com/company/company-name
    - https://linkedin.com/company/company-name/
    - linkedin.com/company/company-name
    """
    if not url:
        raise ValueError("LinkedIn company URL cannot be empty")

    # Remove whitespace
    url = url.strip()

    # Extract company slug from various formats
    patterns = [
        r'(?:https?://)?(?:www\.)?linkedin\.com/company/([^/?\s]+)',
        r'^/company/([^/?\s]+)',
        r'^company/([^/?\s]+)',
    ]

    for pattern in patterns:
        match = re.search(pattern, url, re.IGNORECASE)
        if match:
            company_slug = match.group(1)
            return f"https://www.linkedin.com/company/{company_slug}"

    # If URL doesn't match patterns, assume it's just a company slug
    if '/' not in url and '.' not in url:
        return f"https://www.linkedin.com/company/{url}"

    raise ValueError(f"Invalid LinkedIn company URL: {url}")


def extract_linkedin_username(url: str) -> Optional[str]:
    """Extract just the username from a LinkedIn URL"""
    try:
        normalized = normalize_linkedin_url(url)
        return normalized.split('/in/')[-1]
    except ValueError:
        return None


def extract_company_slug(url: str) -> Optional[str]:
    """Extract just the company slug from a LinkedIn URL"""
    try:
        normalized = normalize_linkedin_company_url(url)
        return normalized.split('/company/')[-1]
    except ValueError:
        return None
